linker.ts

```diff
@@ -1,72 +1,132 @@
+/**
+ * Newer versions of the compiler use hashed names instead of the string values.
+ *
+ * This helper makes the conversion between name and hash at the target length
+ * for replacement.
+ *
+ * @param libraryName
+ */
+function libraryHashPlaceholder (libraryName) {
+  return `$${keccak256(libraryName).slice(0, 34)}$`;
 }
 
-const linkBytecode = function (bytecode, libraries) {
+/**
+ * Used to support legacy library names and the current hash library names,
+ * replaces the names with the provided hex address.
+ *
+ * @param bytecode
+ * @param hexAddress
+ * @param labelName
+ */
+function replacePlaceholderLabelWithHexAddress (bytecode, hexAddress, labelName) {
+  // truncate to 36 characters, why, I assume this is because legacy names
+  // only had a length of 36 characters?
+  const truncatedName = labelName.slice(0, 36);
+
+  const prefix = '__';
+  const postFix = `${Array(37 - truncatedName.length).join('_')}__`;
+  const libLabel = `${prefix}${truncatedName}${postFix}`;
+
+  while (bytecode.indexOf(libLabel) >= 0) {
+    bytecode = bytecode.replace(libLabel, hexAddress);
+  }
+
+  return bytecode;
+}
+
+/**
+ * When using libraries, the resulting bytecode will contain placeholders for
+ * the real addresses of the referenced libraries. These have to be updated,
+ * via a process called linking, before deploying the contract.
+ *
+ * This method provides a simple helper for linking.
+ *
+ * @param {string} bytecode
+ * @param {string} libraries
+ * @returns {string} bytecode
+ */
+function linkBytecode (bytecode, libraries) {
   assert(typeof bytecode === 'string');
   assert(typeof libraries === 'object');
+
   // NOTE: for backwards compatibility support old compiler which didn't use file names
   const librariesComplete = {};
-  for (const libraryName in libraries) {
-    if (typeof libraries[libraryName] === 'object') {
-      // API compatible with the standard JSON i/o
-      for (const lib in libraries[libraryName]) {
-        librariesComplete[lib] = libraries[libraryName][lib];
-        librariesComplete[libraryName + ':' + lib] = libraries[libraryName][lib];
-      }
-    } else {
-      // backwards compatible API for early solc-js versions
-      const parsed = libraryName.match(/^([^:]+):(.+)$/);
-      if (parsed) {
-        librariesComplete[parsed[2]] = libraries[libraryName];
+
+  for (const [libraryName, library] of Object.entries(libraries)) {
+    if (isNil(library)) continue;
+
+    // API compatible with the standard JSON i/o
+    if (isObject(library)) {
+      for (const [key, value] of Object.entries(library)) {
+        librariesComplete[key] = value;
+        librariesComplete[`${libraryName}:${key}`] = value;
       }
-      librariesComplete[libraryName] = libraries[libraryName];
+
+      continue;
     }
+
+    // backwards compatible API for early solc-js versions
+    const parsed = libraryName.match(/^([^:]+):(.+)$/);
+
+    if (!isNil(parsed)) librariesComplete[parsed[2]] = library;
```

Comment:
```suggestion
    if (!isNil(parsed))
      librariesComplete[parsed[2]] = library;
```

Reply:
changed to: 

```js
if (!isNil(parsed)) {
    librariesComplete[parsed[2]] = library;
}
```

Reply:
Good enough :) I mean, I really don't understand why JS linters insist on putting braces there but at least it's still better than having this squashed into a single line, which is extremely unreadable to me.